src/main/java/Duke.java

```diff
@@ -1,273 +1,65 @@
-            displayMessage("Nice! I've marked this task as done:\n" +
-                    tasks.get(listIndex).toString() + "\n");
-        } catch (NumberFormatException e) {
-            displayMessage("Please specify a numerical task index to mark\n");
-        } catch (IndexOutOfBoundsException e) {
-            displayMessage("Please specify a valid index to mark\n");
-        }
-    }
-
-    private void unmarkListItem(String[] tokens) {
-        try {
-            int listIndex = Integer.parseInt(tokens[1]) - 1;
-            tasks.get(listIndex).setStatus(" ");
-            displayMessage("OK, I've marked this task as not done yet:\n" +
-                    tasks.get(listIndex).toString() + "\n");
-        } catch (NumberFormatException e) {
-            displayMessage("Please specify a numerical task index to unmark\n");
-        } catch (IndexOutOfBoundsException e) {
-            displayMessage("Please specify a valid index to unmark\n");
-        }
-    }
-
-    private void deleteItem(String[] tokens) throws DukeException {
-        if (tokens.length != 2) {
-            throw new DukeException("please specify delete command as delete [list index]");
-        } else if (tasks.size() == 0) {
-            throw new DukeException("Task list is empty");
-        }
-        try {
-            int listIndex = Integer.parseInt(tokens[1]) - 1;
-            Task removed = tasks.remove(listIndex);
-            displayMessage("Noted. I've removed this task:\n" +
-                    removed.toString() +
-                    "\nNow you have " + tasks.size() + " tasks in the list\n");
-        } catch (NumberFormatException e) {
-            displayMessage("please specify a valid number to delete entry\n");
-        } catch (IndexOutOfBoundsException e) {
-            displayMessage("please specify a valid index to delete\n");
-        }
-    }
-
-    private void updateData() {
-        String path = "src/data/duke.txt";
-        File file = new File(path);
-        file.getParentFile().mkdirs();
-        try {
-            FileWriter fw = new FileWriter(file);
-            StringBuilder sb = new StringBuilder();
-            for (Task task : tasks) {
-                sb.append(task.asTokens()).append('\n');
-            }
-            fw.write(sb.toString());
-            fw.close();
-        } catch (IOException e) {
-            System.out.println(e.getMessage());
-        }
-    }
-
-    public static void main(String[] args) {
-        Duke duke = new Duke();
-        duke.init();
-        Scanner sc = new Scanner(System.in);
-
+    private void run() {
+        ui.welcomeMessage();
         while (true) {
-            String input = sc.nextLine();
-            String[] tokens = input.split(" ");
-            if (tokens.length == 1 && input.equals("bye")) {
-                duke.exit();
+            String[] tokens = parser.parseUserInput();
+            if (tokens.length == 1 && tokens[0].equals("bye")) {
+                exit();
```

Comment:
Love the idea of using the parser to break down the sentences! :) but could also consider using a switch to improve readability, but this works well too! :)